model/riscv_vmem.sail

```diff
@@ -0,0 +1,854 @@
+                                          false);
+            match mv {
+              MemValue(_) => {
+                add_to_TLB(asid, vAddr, pAddr, pte', pteAddr, level, global,
+                           sv_params.vpn_size_bits,
+                           pagesize_bits);
+                TR_Address(pAddr, ext_ptw)
+              },
+              MemException(e) =>
+                TR_Failure(PTW_Access(), ext_ptw)
+            }
+          }
+        }
+      }
+    }
+}
+
+// PRIVATE
+function translate(sv_params : SV_Params,
+                   asid      : bits(16),
+                   ptb       : bits(64),
+                   vAddr_arg : bits(64),
+                   ac        : AccessType(ext_access_type),
+                   priv      : Privilege,
+                   mxr       : bool,
+                   do_sum    : bool,
+                   ext_ptw   : ext_ptw)
+                  -> TR_Result(bits(64), PTW_Error) = {
+  let va_mask : bits(64) = zero_extend(ones(sv_params.va_size_bits));
+  let vAddr              = (vAddr_arg & va_mask);
+
+  // On first reading, assume lookup_TLB returns None(), since TLBs
+  // are not part of RISC-V archticture spec (see TLB_NOTE above)
+  match lookup_TLB(asid, vAddr) {
+    Some(index, ent) => translate_TLB_hit(sv_params, asid, ptb, vAddr, ac, priv,
+                                          mxr, do_sum, ext_ptw, index, ent),
+    None()           => translate_TLB_miss(sv_params, asid, ptb, vAddr, ac, priv,
+                                           mxr, do_sum, ext_ptw)
+  }
+}
+
+// Top-level addr-translation function
+// PUBLIC: invoked from instr-fetch and load/store/amo
+function translateAddr(vAddr   : xlenbits,
+                       ac      : AccessType(ext_access_type))
+                      -> TR_Result(xlenbits, ExceptionType) = {
+  // Internally the vmem code works with 64-bit values, whether xlen==32 or xlen==64
+  // This 'extend' is a no-op when xlen==64 and extends when xlen==32
+  let vAddr_64b : bits(64) = zero_extend(vAddr);
+  // Effective privilege takes into account mstatus.PRV, mstatus.MPP
+  // See riscv_sys_regs.sail for effectivePrivilege() and cur_privilege
+  let effPriv   : Privilege = effectivePrivilege(ac, mstatus, cur_privilege);
+  let mode      : SATPMode = translationMode(effPriv);
+  // PTW extensions (non-standard): initialize the PTW extension state
+  let ext_ptw   : ext_ptw = init_ext_ptw;
+  let (valid_va, sv_params) : (bool, SV_Params) = match mode {
+    Sbare => return TR_Address(vAddr, ext_ptw),
+    Sv32  => (true,                                 sv32_params),
+    Sv39  => (is_valid_vAddr(sv39_params, vAddr_64b), sv39_params),
+    Sv48  => (is_valid_vAddr(sv48_params, vAddr_64b), sv48_params),
+    // Sv57 => (is_valid_vAddr(sv57_params, vAddr_64b), sv57_params),    // TODO
+  };
+  if not(valid_va) then
+    TR_Failure(translationException(ac, PTW_Invalid_Addr()), ext_ptw)
+  else {
+    let mxr    : bool     = mstatus.MXR() == 0b1;
```

Comment:
Delete all these type annotations.

Reply:
Ok (partial). I’ve deleted them for the obvious boolean exprs, but not for the others, where I think they serve as useful documentation.